### ISSUE TITLE
xSQLServerFirewall: when installing 2 instances firewall rule is created just for the first instance (different SQL server versions)

### ISSUE THREAD

**Issue by Radamonas** [PowerShell/xSQLServer#461]
**Details of the scenario you try and problem that is occurring:**
xSQLServerFirewall - module
Installing two intances, different versions (2008R2 and 2016). The first one is gets firewall rule created, while the second one is skipped and set step is not executed.
Instances have different feature sets, different names, different install paths and different ports defined.

Once the rule is removed which was created for the first instance and the script was ran for the second time the new firewall rule was created and instance became available for remote connection. So the settings provided for xSQLServerFirewall were valid. After reading removed instance configuration to DSC script, new firewall rule was not created for readded instance. Seems that problem is with test function.

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
Windows2012
PS 5.1

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
Version: 6.0.0.0 (master)
xSQLServerFirewall

**Comment by johlju** [comment]
@Radamonas Could you please provide the configuration you used (change any sensitive information) ? It would help solving this.

**Comment by Radamonas** [comment]
Hi,

Here comes the setup details used for firewall. I also noticed, that if _SQL Server Browser_ is not deleted from Windows Firewall the instance name, aka, **SQL Server Database Engine instance INSTANCE1** called role is not created. 

**NODE SETUP**
 AllNodes = @(
        @{
            NodeName = '*'
            [...]		
        },
        @{
            NodeName = 'SERVER'
            SQLInstances = @(
                @{
                    Instance = 'INSTANCE1'
                    Features = 'SQLENGINE,SSMS,ADV_SSMS'
                    SecurityMode = 'SQL'
                    InstallSubFolder = $SQL2008RFolder
                    Collation = 'Latin1_General_CI_AS'
                    TCPPort = 24212
                },
                 @{
                    Instance = 'INSTANCE2'
                    Features = 'SQLENGINE,IS'
                    SecurityMode = 'SQL'
                    InstallSubFolder = $SQL2016Folder
                    Collation = 'Latin1_General_BIN'
                    TCPPort = 24213
                }
            )
        }
    )

**EXECUTION SETUP**
foreach($sqlInstance in $Node.SQLInstances)
{
    xSqlServerSetup $sqlInstance.Instance
    {
		[..]
    }
    xSqlServerFirewall $sqlInstance.Instance
    {
        DependsOn = @("[xSqlServerSetup]$($sqlInstance.Instance)")         
        SourcePath = Join-Path $Node.SourceLocalPath $sqlInstance.InstallSubFolder
        InstanceName = $sqlInstance.Instance
        Features = $sqlInstance.Features
        Ensure = 'Present'
        PsDscRunAsCredential = $Node.InstallerAccount
    }
	[..]
}

**EXECUTION OUTPUT**
VERBOSE: [SERVER]: LCM:  [ Start  Resource ]  [[xSQLServerFirewall]INSTANCE1]
VERBOSE: [SERVER]: LCM:  [ Start  Test     ]  [[xSQLServerFirewall]INSTANCE1]
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] 2017-04-03_07-27-18: Using path:
C:\Install\SQLServer2008R2_Developer_with_SP3\setup.exe
VERBOSE: [SERVER]: LCM:  [ End    Test     ]  [[xSQLServerFirewall]INSTANCE1]  in 3.9490 seconds.
VERBOSE: [SERVER]: LCM:  [ Start  Set      ]  [[xSQLServerFirewall]INSTANCE1]
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] 2017-04-03_07-27-21: Using path:
C:\Install\SQLServer2008R2_Developer_with_SP3\setup.exe
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] 2017-04-03_07-27-21: Using path:
C:\Install\SQLServer2008R2_Developer_with_SP3\setup.exe
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE1] 2017-04-03_07-27-23: Using path:
C:\Install\SQLServer2008R2_Developer_with_SP3\setup.exe
VERBOSE: [SERVER]: LCM:  [ End    Set      ]  [[xSQLServerFirewall]INSTANCE1]  in 2.6980 seconds.
VERBOSE: [SERVER]: LCM:  [ End    Resource ]  [[xSQLServerFirewall]INSTANCE1]
[...]
VERBOSE: [SERVER]: LCM:  [ Start  Resource ]  [[xSQLServerFirewall]INSTANCE2]
VERBOSE: [SERVER]: LCM:  **[ Start  Test     ]**  [[xSQLServerFirewall]INSTANCE2]
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE2] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE2] Importing the module
MSFT_xSQLServerFirewall in force mode.
VERBOSE: [SERVER]:                            [[xSQLServerFirewall]INSTANCE2] 2017-04-03_07-27-27: Using path:
C:\Install\SqlServer2016_Developer_SP1_x64\setup.exe
VERBOSE: [SERVER]: LCM:  **[ End    Test     ]**  [[xSQLServerFirewall]INSTANCE2]  in 1.1350 seconds.
VERBOSE: [SERVER]: LCM:  **[ Skip   Set      ]**  [[xSQLServerFirewall]INSTANCE2]
VERBOSE: [SERVER]: LCM:  [ End    Resource ]  [[xSQLServerFirewall]INSTANCE2]

**Comment by johlju** [comment]
Thanks for the hint about the SQL Server Browser rule!

Regarding that SQL Server Browser rule must be deleted for the SQL Server Database Engine instance INSTANCE1 rule to be created. I found that bug. Same will happen if reversed.

I _think_ because SQL Server Browser exist the second instance is tested okay, due to the bug above.

**Comment by johlju** [comment]
@Radamonas Would you mind helping me out an test the code in PR #479 and see if it fixes your problem? You find the updated resource in my working branch here https://github.com/johlju/xSQLServer/tree/fix-issue-%23461

**Comment by Radamonas** [comment]
Hi,

Definitely, I will test the fix. I will update you tomorrow about the
output of my testing.

Have a nice Sunday.

2017 bal. 9 12:47 "Johan Ljunggren" <notifications@github.com> rašė:

**Comment by Radamonas** [comment]
Hi,

I've checked your changes on my dev enviroment.

Test cases:
* deleted all SQL related firewall rules
* deleted one SQL instance related rule
* deleted all SQL instance related rules (Browser was left)
* deleted just SQL browser related rule.

All tests were **successful**, all rules were recreated after rerunning the DSC.

Before this fix I haven't noticed that _SQL Server Integration Services Application_ rule was not created, however now it is reappearing after each run, so this issue was also resolved with the fix.

Kudos to @johlju
